File: test/browser_e2e.py
from pathlib import Path
from playwright.sync_api import sync_playwright


BASE_URL = "http://127.0.0.1:8787"
RUNTIME = Path(__file__).resolve().parents[1] / "runtime"
ASSETS = Path(__file__).resolve().parents[1] / "assets"


def prepare_and_approve(page):
    page.goto(BASE_URL)
    page.wait_for_load_state("networkidle")
    page.get_by_role("button", name="Prepare exact artifact").click()
    page.wait_for_selector("#paper:not([hidden])", timeout=30_000)
    phrase = page.locator("#required-phrase").inner_text()
    assert phrase.startswith("APPROVE ")
    page.locator("#approval-phrase").fill(phrase)
    page.locator("#attest-recipient").check()
    page.locator("#attest-authority").check()
    page.get_by_role("button", name="Issue one-shot approval").click()
    page.locator("#send-button:enabled").wait_for(timeout=10_000)


def main():
    RUNTIME.mkdir(parents=True, exist_ok=True)
    ASSETS.mkdir(parents=True, exist_ok=True)
    console_errors = []
    with sync_playwright() as p:
        browser = p.chromium.launch(channel="chrome", headless=True)
        context = browser.new_context(viewport={"width": 1440, "height": 1050}, device_scale_factor=1)
        page = context.new_page()
        page.on("console", lambda message: console_errors.append(message.text) if message.type == "error" else None)
        page.on("pageerror", lambda error: console_errors.append(str(error)))

        prepare_and_approve(page)
        assert "Mutual" in page.locator("#doc-title").inner_text()
        assert len(page.locator("#artifact-hash").inner_text()) == 64
        assert "HTML PROOF / DEMO" in page.locator("#artifact-mode").inner_text()
        page.get_by_role("button", name="Dispatch approved artifact").click()
        page.wait_for_function("() => document.querySelector('#event-log')?.innerText.includes('sent no email')")
        assert page.locator("#send-button").is_disabled()
        page.locator(".workspace").screenshot(path=str(ASSETS / "signgate-workspace.png"))
        page.screenshot(path=str(RUNTIME / "e2e-desktop.png"), full_page=True)

        prepare_and_approve(page)
        page.get_by_role("button", name="Change recipient, then try ↯").click()
        page.wait_for_function("() => document.querySelector('#event-log')?.innerText.includes('EXPECTED HARD BLOCK')")
        assert "APPROVAL_BINDING_BROKEN" in page.locator("#event-log").inner_text()
        assert "ZERO PROVIDER CALLS" in page.locator("#gate-result").inner_text()
        page.locator(".workspace").screenshot(path=str(ASSETS / "signgate-tamper-proof.png"))
        assert page.locator("#audit-valid").inner_text() == "HASH CHAIN VALID"

        overflow = page.evaluate("document.documentElement.scrollWidth - document.documentElement.clientWidth")
        assert overflow <= 1, f"desktop horizontal overflow: {overflow}px"

        mobile = browser.new_context(viewport={"width": 390, "height": 844}, device_scale_factor=1)
        mobile_page = mobile.new_page()
        mobile_page.goto(BASE_URL)
        mobile_page.wait_for_load_state("networkidle")
        mobile_overflow = mobile_page.evaluate("document.documentElement.scrollWidth - document.documentElement.clientWidth")
        assert mobile_overflow <= 1, f"mobile horizontal overflow: {mobile_overflow}px"
        mobile_page.screenshot(path=str(RUNTIME / "e2e-mobile.png"), full_page=True)
        mobile_page.screenshot(path=str(ASSETS / "signgate-mobile.png"), full_page=False)

        mobile.close()
        context.close()
        browser.close()

    unexpected_errors = [message for message in console_errors if "409 (Conflict)" not in message]
    assert not unexpected_errors, f"browser console errors: {unexpected_errors}"
    print("E2E PASS: prepare → approve → transparent simulation")
    print("E2E PASS: post-approval recipient mutation → hard block")
    print("E2E PASS: desktop/mobile layouts have no horizontal overflow")
    print(f"Screenshots: {RUNTIME / 'e2e-desktop.png'}, {RUNTIME / 'e2e-mobile.png'}")


if __name__ == "__main__":
    main()

File: test/browser-e2e.mjs
import assert from 'node:assert/strict';
import { mkdir } from 'node:fs/promises';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { chromium } from 'playwright';

const root = resolve(dirname(fileURLToPath(import.meta.url)), '..');
const baseUrl = process.env.SIGNGATE_BASE_URL ?? 'http://127.0.0.1:8787';
const runtimeDir = resolve(root, 'runtime');
const assetsDir = resolve(root, 'assets');

async function prepareAndApprove(page) {
  await page.goto(baseUrl);
  await page.waitForLoadState('networkidle');
  await page.locator('#provenance-particles canvas').waitFor({ timeout: 10_000 });
  await page.getByRole('button', { name: 'Prepare exact artifact' }).click();
  await page.locator('#paper:not([hidden])').waitFor({ timeout: 30_000 });
  await page.locator('[data-route-stage="approve"].is-active').waitFor({ timeout: 10_000 });
  const phrase = await page.locator('#required-phrase').innerText();
  assert.match(phrase, /^APPROVE /);
  await page.locator('#approval-phrase').fill(phrase);
  await page.locator('#attest-recipient').check();
  await page.locator('#attest-authority').check();
  await page.getByRole('button', { name: 'Issue one-shot approval' }).click();
  await page.locator('#send-button:enabled').waitFor({ timeout: 10_000 });
  await page.locator('[data-route-stage="send"].is-active').waitFor({ timeout: 10_000 });
  assert.equal(await page.locator('#seal-label').innerText(), 'APPROVED');
  await page.waitForFunction(() => getComputedStyle(document.querySelector('#scan-beam')).opacity === '0');
}

await mkdir(runtimeDir, { recursive: true });
await mkdir(assetsDir, { recursive: true });
const consoleErrors = [];
const browser = await chromium.launch({ channel: 'chrome', headless: true });

try {
  const context = await browser.newContext({ viewport: { width: 1440, height: 1050 }, deviceScaleFactor: 1 });
  const page = await context.newPage();
  page.on('console', (message) => {
    if (message.type() === 'error') consoleErrors.push(message.text());
  });
  page.on('pageerror', (error) => consoleErrors.push(error.message));

  await prepareAndApprove(page);
  assert.match(await page.locator('#doc-title').innerText(), /Mutual/);
  assert.equal((await page.locator('#artifact-hash').innerText()).length, 64);
  assert.match(await page.locator('#artifact-mode').innerText(), /HTML PROOF \/ DEMO/);
  await page.getByRole('button', { name: 'Dispatch approved artifact' }).click();
  await page.waitForFunction(() => document.querySelector('#event-log')?.innerText.includes('sent no email'));
  assert.equal(await page.locator('#send-button').isDisabled(), true);
  assert.equal(await page.locator('#seal-label').innerText(), 'CONSUMED');
  assert.match(await page.locator('[data-route-stage="proof"]').getAttribute('class'), /is-active/);
  assert.ok(await page.locator('#prepare-button').evaluate((element) => element.getBoundingClientRect().height >= 44));
  await page.locator('.workspace').screenshot({ path: resolve(assetsDir, 'signgate-workspace.png') });
  await page.screenshot({ path: resolve(runtimeDir, 'e2e-desktop.png'), fullPage: true });

  await prepareAndApprove(page);
  await page.getByRole('button', { name: 'Change recipient, then try ↯' }).click();
  await page.waitForFunction(() => document.querySelector('#event-log')?.innerText.includes('EXPECTED HARD BLOCK'));
  assert.match(await page.locator('#event-log').innerText(), /APPROVAL_BINDING_BROKEN/);
  assert.match(await page.locator('#gate-result').innerText(), /ZERO PROVIDER CALLS/);
  assert.match(await page.locator('#provenance-board').getAttribute('class'), /is-alert/);
  await page.locator('.workspace').screenshot({ path: resolve(assetsDir, 'signgate-tamper-proof.png') });
  assert.equal(await page.locator('#audit-valid').innerText(), 'HASH CHAIN VALID');
  assert.ok(await page.evaluate(() => document.documentElement.scrollWidth - document.documentElement.clientWidth) <= 1);

  const mobile = await browser.newContext({ viewport: { width: 390, height: 844 }, deviceScaleFactor: 1 });
  const mobilePage = await mobile.newPage();
  mobilePage.on('console', (message) => {
    if (message.type() === 'error') consoleErrors.push(message.text());
  });
  mobilePage.on('pageerror', (error) => consoleErrors.push(error.message));
  await mobilePage.goto(baseUrl);
  await mobilePage.waitForLoadState('networkidle');
  assert.ok(await mobilePage.evaluate(() => document.documentElement.scrollWidth - document.documentElement.clientWidth) <= 1);
  await mobilePage.screenshot({ path: resolve(runtimeDir, 'e2e-mobile.png'), fullPage: true });
  await mobilePage.screenshot({ path: resolve(assetsDir, 'signgate-mobile.png'), fullPage: false });

  const reduced = await browser.newContext({
    viewport: { width: 1280, height: 800 },
    deviceScaleFactor: 1,
    reducedMotion: 'reduce',
  });
  const reducedPage = await reduced.newPage();
  reducedPage.on('console', (message) => {
    if (message.type() === 'error') consoleErrors.push(message.text());
  });
  reducedPage.on('pageerror', (error) => consoleErrors.push(error.message));
  await reducedPage.goto(baseUrl);
  await reducedPage.waitForLoadState('networkidle');
  assert.equal(await reducedPage.locator('#provenance-particles').evaluate((element) => getComputedStyle(element).display), 'none');
  assert.equal(await reducedPage.locator('#route-packet').evaluate((element) => getComputedStyle(element).display), 'none');
  assert.ok(await reducedPage.evaluate(() => document.documentElement.scrollWidth - document.documentElement.clientWidth) <= 1);

  await reduced.close();
  await mobile.close();
  await context.close();
} finally {
  await browser.close();
}

const unexpectedErrors = consoleErrors.filter((message) => !message.includes('409 (Conflict)'));
assert.deepEqual(unexpectedErrors, [], `browser console errors: ${unexpectedErrors.join(' | ')}`);
console.log('E2E PASS: prepare → approve → transparent simulation');
console.log('E2E PASS: post-approval recipient mutation → hard block');
console.log('E2E PASS: desktop/mobile layouts have no horizontal overflow');
console.log('E2E PASS: state-linked motion loads; reduced-motion disables ambient effects');
console.log(`Screenshots: ${resolve(runtimeDir, 'e2e-desktop.png')}, ${resolve(runtimeDir, 'e2e-mobile.png')}`);
